src/data/nat/digits.lean

```diff
@@ -0,0 +1,346 @@
+  { dsimp [of_digits],
+    rw dif_neg at w₂,
+    swap,
+    { simp, },
+    rw digits_add b h,
+    { rw L_ih,
+      { simp, },
+      { intros l m, apply w₁, exact list.mem_cons_of_mem _ m, },
+      { split_ifs,
+        { trivial, },
+        { rw [list.last_cons _ h_1] at w₂,
+            convert w₂, }}},
+    { convert w₁ L_hd (list.mem_cons_self _ _), simp, },
+    { by_cases h' : L_tl = [],
+      { rcases h' with rfl,
+        simp at w₂,
+        left,
+        apply nat.pos_of_ne_zero,
+        convert w₂, simp, },
+      { right,
+        apply nat.pos_of_ne_zero,
+        contrapose! w₂,
+        apply digits_zero_of_eq_zero _ w₂,
+        { rw list.last_cons _ h',
+          exact list.last_mem h', },
+        { exact le_of_lt h, }, }, }, },
+end
+
+lemma of_digits_digits (b n : ℕ) : of_digits b (digits b n) = n :=
+begin
+  cases b with b,
+  { cases n with n,
+    { refl, },
+    { change of_digits 0 [n+1] = n+1,
+      dsimp [of_digits],
+      simp, } },
+  cases b with b,
+  { induction n with n,
+    { refl, },
+    { simp [n_ih, add_comm 1], } },
+  apply nat.strong_induction_on n _, clear n,
+  intros n h,
+  cases n,
+  { refl, },
+  { change of_digits (b+2) ((n+1) % (b+2) :: digits (b+2) ((n+1) / (b+2))) = (n+1),
+    dsimp [of_digits],
+    rw h _ (nat.div_lt_self' n b),
+    simp [nat.mod_add_div], },
+end
+
+lemma of_digits_one (L : list ℕ) : of_digits 1 L = L.sum :=
+begin
+  induction L,
+  { refl, },
+  { dsimp [of_digits], simp [list.sum_cons, L_ih], }
+end
+
+-- I don't want to move this to `euclidean_domain.lean`, because `ring` isn't available there.
+lemma dvd_mul_sub_mul {α : Type*} [euclidean_domain α]
+  {k a b x y : α} (hab : k ∣ a - b) (hxy : k ∣ x - y) :
+  k ∣ a * x - b * y :=
+begin
+  cases hab,
+  rw eq_add_of_sub_eq hab_h,
+  cases hxy,
+  rw eq_add_of_sub_eq hxy_h,
+  use k * hab_w * hxy_w + hab_w * y + b * hxy_w,
+  ring,
+end
+
+lemma dvd_of_digits_sub_of_digits {α : Type*} [euclidean_domain α]
+  {a b k : α} (h : k ∣ a - b) (L : list ℕ) :
+  k ∣ ((of_digits a L) - (of_digits b L)) :=
+begin
+  induction L,
+  { change k ∣ 0 - 0, simp, },
+  { dsimp [of_digits],
+    simp only [add_sub_add_left_eq_sub],
+    exact dvd_mul_sub_mul h L_ih, }
+end
+
+lemma of_digits_modeq' (b b' : ℕ) (k : ℕ) (h : b ≡ b' [MOD k]) (L : list ℕ) : (of_digits b L) ≡ (of_digits b' L) [MOD k] :=
+begin
+  induction L,
+  { refl, },
+  { dsimp [of_digits],
+    dsimp [nat.modeq] at *,
+    conv_lhs { rw [nat.add_mod, nat.mul_mod, h, L_ih], },
+    conv_rhs { rw [nat.add_mod, nat.mul_mod], }, }
+end
+
+lemma of_digits_modeq (b k : ℕ) (L : list ℕ) : (of_digits b L) ≡ (of_digits (b % k) L) [MOD k] :=
+of_digits_modeq' b (b % k) k (nat.modeq.symm (nat.modeq.mod_modeq b k)) L
+
+lemma of_digits_mod (b k : ℕ) (L : list ℕ) : (of_digits b L) % k = (of_digits (b % k) L) % k :=
+of_digits_modeq b k L
+
+lemma of_digits_zmodeq' (b b' : ℤ) (k : ℕ) (h : b ≡ b' [ZMOD k]) (L : list ℕ) : (of_digits b L) ≡ (of_digits b' L) [ZMOD k] :=
```

Comment:
Some of your lines are really too long :)

Reply:
Indeed!